File: BOJ/Gold V/BOJ3055.py
#BOJ3055 탈출 20210317
import sys
from collections import deque
input = sys.stdin.readline

def main():
    r,c = map(int, input().rstrip().split())
    li = [list(input().rstrip()) for _ in range(r)]
    water = []
    gosum = ""
    d = ""
    for i in range(r):
        for j in range(c):
            if li[i][j] == "S":
                gosum = [(i,j)]
                li[i][j] = "S0"
            if li[i][j] == "*":
                water.append((i,j))
                li[i][j] = "*0"
            if li[i][j] == "D":
                d = (i,j)
    waterQ = deque(water)
    q = deque(gosum)
    time = 0
    dc = [0, 1, 0, -1]
    dr = [1, 0, -1, 0]
    flag = False
    while q or waterQ:
        if flag: break
        while waterQ:
            wr,wc = waterQ[0]
            if li[wr][wc] != "*"+str(time): break
            waterQ.popleft()
            for i in range(4):
                nwr = wr + dr[i]
                nwc = wc + dc[i]
                if 0 <= nwr < r and 0 <= nwc < c:
                    if li[nwr][nwc] == ".":
                        li[nwr][nwc] = "*"+str(time+1)
                        waterQ.append((nwr,nwc))
        while q:
            sr,sc = q[0]
            if li[sr][sc] != "S"+str(time): break
            q.popleft()
            for i in range(4):
                nsr = sr + dr[i]
                nsc = sc + dc[i]
                if 0 <= nsr < r and 0 <= nsc < c:
                    if li[nsr][nsc] == "D":
                        li[nsr][nsc] = time+1
                        flag = True
                    if li[nsr][nsc] == ".":
                        li[nsr][nsc] = "S"+str(time+1)
                        q.append((nsr,nsc))
        time += 1    
    '''
    for v in li:
        print(v)'''
    print("KAKTUS" if li[d[0]][d[1]] == "D" else li[d[0]][d[1]])
if __name__ == '__main__':
    main()

File: BOJ/Gold V/BOJ3055.java
//BOJ3055 탈출 20210705
import java.io.*;
import java.util.*;

public class BOJ3055 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        StringTokenizer stk = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(stk.nextToken());
        int m = Integer.parseInt(stk.nextToken());

        String[][] board = new String[n][m];
        for(int i = 0; i < n; i++){
            board[i] = br.readLine().split("");
        }


        Queue<int[]> gosumLoc = new LinkedList<>();
        Queue<int[]> waterQ = new LinkedList<>();
        int[] cave = new int[2];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j ++){
                if(board[i][j].equals("S")){
                    gosumLoc.add(new int[]{i,j});
                    board[i][j] = "S0";
                }
                else if(board[i][j].equals("D")){
                    cave = new int[]{i,j};
                }
                else if(board[i][j].equals("*")){
                   waterQ.add(new int[]{i,j});
                   board[i][j] = "*0";
                }
            }
        }

        int[] dr = {-1, 0, 1, 0};
        int[] dc = {0, 1, 0, -1};
        int time = 0;
        boolean flag = false;

        while(!gosumLoc.isEmpty() || !waterQ.isEmpty()){
            if(flag) break;
            while(!waterQ.isEmpty()){
                int[] pos = waterQ.peek();
                int r = pos[0], c = pos[1];
                if(!board[r][c].equals("*" + time)) break;
                waterQ.poll();
                for(int i = 0; i < 4; i++){
                    int nr = r + dr[i];
                    int nc = c + dc[i];
                    if((0 <= nr && nr < n) && (0 <= nc && nc < m)){
                        if(board[nr][nc].equals(".")){
                            board[nr][nc] = "*" + (time + 1);
                            waterQ.add(new int[]{nr, nc});
                        }
                    }
                }
            }

            while(!gosumLoc.isEmpty()){
                int[] pos = gosumLoc.peek();
                int r = pos[0], c = pos[1];
                if(!board[r][c].equals("S" + time)) break;
                gosumLoc.poll();

                for(int i = 0; i < 4; i++){
                    int nr = r + dr[i];
                    int nc = c + dc[i];
                    if((0 <= nr && nr < n) && (0 <= nc && nc < m)){
                        if(board[nr][nc].equals("D")){
                            board[nr][nc] = Integer.toString(time+1);
                            flag = true;
                        }
                        if(board[nr][nc].equals(".")){
                            board[nr][nc] = "S" + (time + 1);
                            gosumLoc.add(new int[]{nr,nc});
                        }
                    }
                }
            }
            time++;
        }
        String ret = board[cave[0]][cave[1]];
        if(ret.equals("D")){
            bw.write("KAKTUS\n");
        }else{
            bw.write(ret + "\n");
        }

        bw.flush();
        bw.close();
    }
}